flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/calcite/FlinkCalciteSqlValidator.java

```diff
@@ -125,4 +126,14 @@ public void validateColumnListParams(
         // this makes it possible to ignore them in the validator and fall back to regular row types
         // see also SqlFunction#deriveType
     }
+
+    @Override
+    public void validateInsert(SqlInsert insert) {
+        super.validateInsert(insert);
+
+        // We don't support UPSERT INTO semantics (see FLINK-22942).
```

Comment:
Maybe it would be better to open a JIRA issue for supporting UPSERT (as a placeholder, simply stating that the semantics need to be defined first) and then link to that ticket from here rather than the one that causes us to disallow it? What do you think?

Reply:
https://issues.apache.org/jira/browse/FLINK-24225